Enable ARM SVE Support on Linux

mohakgoel1 opened the pull request:
**Description**

This PR enables Scalable Vector Extension (SVE) support for RocksDB on ARM-based Linux platforms. It introduces the necessary enablement logic while preserving existing behavior on non-SVE and non-Linux systems.

**Key Changes**

- Added SVE enablement logic for ARM-based Linux systems
- Verified build compatibility and functional correctness with existing configurations
- Ensured safe fallback to existing implementations when SVE is not available
- Prepared the groundwork for future SVE-based performance optimizations

**Compatibility & Safety**

- No behavior change on platforms without SVE support
- No impact on x86 or non-ARM architectures
- Linux-only enablement to avoid affecting unsupported environments

**Testing**

- Built and tested on ARM Linux systems

**Benchmark Description**

The YCSB (Yahoo! Cloud Serving Benchmark) framework was used to evaluate the performance of RocksDB on an ARM-based platform. Both load and run phases were executed across standard YCSB workloads (A–F). The experiments compare performance with and without SVE enabled, measuring execution time in milliseconds. The results below show the impact of SVE on RocksDB performance for different workload patterns.

The values represent runtime (ms), and a higher ratio indicates better performance with SVE enabled. Please have a look.

### YCSB Load Results

| Workload | No SVE (ms) | With SVE (ms) | Ratio (No SVE / With SVE) |
|----------|-------------|---------------|----------------------------|
| A        | 2881        | 2880          | 1.00 |
| B        | 2875        | 2866          | 1.00 |
| C        | 2871        | 2867          | 1.00 |
| D        | 2872        | 2869          | 1.00 |
| E        | 2875        | 2863          | 1.00 |
| F        | 2883        | 2881          | 1.00 |


### YCSB Run Results

| Workload | No SVE (ms) | With SVE (ms) | Ratio (No SVE / With SVE) |
|----------|-------------|---------------|----------------------------|
| A        | 4298        | 4249          | 1.01 |
| B        | 3762        | 3713          | 1.01 |
| C        | 3739        | 3700          | 1.01 |
| D        | 3768        | 3709          | 1.02 |
| E        | 4822        | 4759          | 1.01 |
| F        | 4299        | 4248          | 1.01 |


**This PR is a joint contribution by:**

- Abhishek Jain(@abhijain1204fujitsu)
- Mohak Goel(@mohakgoel1)

mohakgoel1 committed:
SVE Enablement

mohakgoel1 committed:
Fixed override issue

mohakgoel1 committed:
Override issue fix

mohakgoel1 committed:
Missing endif fix

mohakgoel1 commented:
Hi @adamretter @jaykorean — just a gentle reminder about this PR when you have some time.
It adds ARM SVE support for RocksDB on Linux and I’d appreciate your feedback.  
Thank you!

xingbowang commented:
Could you add some benchmark results with these changes on arm hosts?

mohakgoel1 commented:
> Could you add some benchmark results with these changes on arm hosts?

Hi @xingbowang,
Thank you for the quick reply!
I’ve updated the PR description with the benchmark results. Please let me know if anything else is needed.

xingbowang commented:
Why there is very little to no performance gain? Was it due to the test hardware, or compiler, or we have to rewrite some part of the code to fully extract the performance gain?

mohakgoel1 committed:
Moved changes to build_detect_platform

mohakgoel1 committed:
Added runtime check for SVE

mohakgoel1 committed:
Added runtime check for SVE

mohakgoel1 committed:
Moved changes to build_detect_platform file

mohakgoel1 committed:
cached sve runtime detection

mohakgoel1 commented on Makefile:
```diff
@@ -279,11 +279,26 @@ endif
-CXXFLAGS += -march=armv8-a+crc+crypto
-CFLAGS += -march=armv8-a+crc+crypto
-ARMCRC_SOURCE=1
+ifneq (,$(shell $(CXX) -fsyntax-only -march=armv8.2-a+sve+crc+crypto -xc /dev/null 2>&1))
```
@pdillinger, Just a gentle reminder for this PR. It would be great if you could review it when you get a chance.

Please let me know if you need anything from my side. Thanks!

mohakgoel1 commented:
Hi @xingbowang, @pdillinger
I’ve incorporated the suggested changes and added an additional runtime check for SVE detection, ensuring all relevant cases are now covered. Thank you for your valuable suggestion.
Whenever you have some time, I would appreciate it if you could take a look at this PR.
